main.go

```diff
@@ -3,84 +3,143 @@ package main
+	recursive       bool
+}
 
-	if !atLeastOneTrue {
-		fmt.Fprint(os.Stderr, "err: at least one event should be specified\n")
-		os.Exit(1)
-	}
+func (opt *watcherOptions) print() {
+	//TODO: colorize the option with emoji
+	fmt.Println("watcher version 0.1.0")
+	fmt.Println("path: ", opt.path)
+	fmt.Println("commands: ", opt.commands)
+	fmt.Println("events: ", opt.registredEvents)
+	fmt.Println("recursive: ", opt.recursive)
+}
 
-	_, err := os.Stat(path)
+func validateAndParseFlags(
+	commands string,
+	path string,
+	events string,
+	recursive bool,
+) (opt watcherOptions) {
+	// a function to validate the flag and build the watcherOptions struct
+	fileInfo, err := os.Stat(path)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "err: %v\n", err)
 		os.Exit(1)
+	} else {
+		opt.path = path
 	}
 
-	watcher, _ := fsnotify.NewWatcher()
-	err = addSubdirectories(path, watcher)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "err: %v\n", err)
+	if recursive && !fileInfo.IsDir() {
+		fmt.Fprintf(os.Stderr, "err: %v\n", "recursive flag can only be used with directories")
 		os.Exit(1)
+	} else {
+		opt.recursive = recursive
 	}
 
-	eventTime := time.Now()
+	// parse events
+	if events == "all" {
+		opt.registredEvents = []fsnotify.Op{
+			fsnotify.Write,
+			fsnotify.Create,
+			fsnotify.Chmod,
+			fsnotify.Remove,
+			fsnotify.Rename,
+		}
+	} else {
+		events = strings.ToLower(events)
+		eventsList := strings.Split(events, ",")
+		for _, event := range eventsList {
+			event = strings.TrimSpace(event)
+			switch event {
+			case "write":
+				opt.registredEvents = append(opt.registredEvents, fsnotify.Write)
+			case "create":
+				opt.registredEvents = append(opt.registredEvents, fsnotify.Create)
+			case "chmod":
+				opt.registredEvents = append(opt.registredEvents, fsnotify.Chmod)
+			case "remove":
+				opt.registredEvents = append(opt.registredEvents, fsnotify.Remove)
+			case "rename":
+				opt.registredEvents = append(opt.registredEvents, fsnotify.Rename)
+			default:
+				fmt.Fprintf(os.Stderr, "err: %v%s\n", "invalid event: ", event)
+				os.Exit(1)
+			}
+		}
+	}
+	opt.commands = parseCommands(commands)
+	return
+}
 
+func watchEvents(watcher *fsnotify.Watcher, options watcherOptions) {
 	for {
 		select {
-		case event := <-watcher.Events:
-			for k, ok := range m {
-				if event.Has(k) && ok && time.Since(eventTime) > time.Millisecond*500 {
-					cmds := parseCommands(command)
-					for _, c := range cmds {
-						c.Stderr = os.Stderr
-						c.Stdout = os.Stdout
-
-						err = c.Start()
-						if err != nil {
-							fmt.Fprintf(os.Stderr, "err: %v\n", err)
-							os.Exit(1)
-						}
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+			for _, op := range options.registredEvents {
+				if event.Has(op) {
+					log.Printf("%v on %v => executing command %v\n", event.Op, event.Name, options.commands)
+					for _, cmd := range options.commands {
+						cmd.Start()
+						cmd.Stdout = os.Stdout
+						cmd.Stderr = os.Stderr
```

Comment:
the output of the commands are not visible on `os.Stdout` or `os.Stderr`,
that's because you are starting the command before updating the `cmd.Stdout` and `cmd.Stderr`